3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/VeroAdapter.scala

```diff
@@ -0,0 +1,167 @@
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  private def payloadBodyToEvent(body_json: String, payload: CollectorPayload): Validated[RawEvent] =
+    try {
+      val parsed    = parse(body_json)
+      val eventType = (parsed \ "type").extractOpt[String].orElse(Some("user_updated"))
+      val formattedEvent = eventType match {
+        case Some("user_updated") => reformatParameters(parsed)
+        case _ => {
+          cleanupJsonEventValues(
+            parsed,
+            ("type", eventType.get).some,
+            eventType.get.concat("_at")
+          )
+        }
+      }
+
+      val reformattedEvent = reformatParameters(formattedEvent)
+
+      lookupSchema(eventType, VendorName, EventSchemaMap) map { schema =>
+        RawEvent(
+          api = payload.api,
+          parameters = toUnstructEventParams(
+            TrackerVersion,
+            toMap(payload.querystring),
+            schema,
+            reformattedEvent,
+            "srv"
+          ),
+          contentType = payload.contentType,
+          source      = payload.source,
+          context     = payload.context
+        )
+      }
+
+    } catch {
+      case e: JsonParseException => {
+        val exception = JU.stripInstanceEtc(e.toString).orNull
+        s"$VendorName event failed to parse into JSON: [$exception]".failureNel
+      }
+    }
+
+  /**
+   * Converts a CollectorPayload instance into raw events.
+   * A Vero API payload only contains a single event.
+   * We expect the type parameter to match the supported events, else
+   * we have an unsupported event type.
+   *
+   * @param payload The CollectorPayload containing one or more
+   *        raw events as collected by a Snowplow collector
+   * @param resolver (implicit) The Iglu resolver used for
+   *        schema lookup and validation. Not used
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents =
+    (payload.body, payload.contentType) match {
+      case (None, _) => s"Request body is empty: no $VendorName event to process".failureNel
+      case (Some(body), _) => {
+        val event = payloadBodyToEvent(body, payload)
+        rawEventsListProcessor(List(event))
+      }
+    }
+
+  /**
+   * Returns an updated Vero event JSON where
+   * the "action" field has been removed
+   * and "triggered_at" fields' values have been converted
+   *
+   * @param json The event JSON which we need to
+   *        update values for
+   * @return the updated JSON with updated fields and values
+   */
+  private[registry] def reformatParameters(json: JValue): JValue = {
+
+    def toStringField(value: Long): JString = {
+      val dt: DateTime = new DateTime(value)
+      JString(JsonSchemaDateTimeFormat.print(dt))
+    }
```

Comment:
can this fail? if so the result should ideally be a `Validation[E, JString]`

Reply:
this function was "borrowed" from the `cleanupJsonEventValues` method as is
the function is called only when it matches the case of
```scala
case ("triggered_at", JInt(x)")
```
so I'm not sure if any possible errors could arise if some integer values exist for that field to be converted
let me know what you guys think

Reply:
I think we're good